make ClientM  a Reader

chkl opened the pull request:
As discussed in https://github.com/haskell-servant/servant/issues/563

This would change the definition of clientM from
```haskell
type ClientM = ExceptT ServantError IO
```

to
```haskell
newtype ClientM a = ClientM { runClientM' :: ReaderT ClientEnv (ExceptT ServantError IO) a }
```

where ClientEnv is just a record of `Manager` and `BaseUrl`

I added `mtl` as a dependency,  but I am not sure about the version boundaries.

jkarni commented:
CI is failing for ghc 7.8.4 (run `stack test --stack-yaml=stack-ghc-7.8.4.yaml` to test it locally). I think because of the version of `mtl` picked.

jkarni commented on doc/tutorial/Client.lhs:
````diff
@@ -120,17 +113,17 @@ data BaseUrl = BaseUrl
 That's it. Let's now write some code that uses our client functions.
 
 ``` haskell
````
This is great. All the signatures have become a lot cleaner!

jkarni commented on servant-client/src/Servant/Common/Req.hs:
```diff
@@ -149,20 +152,39 @@ parseRequest url = liftM disableStatusCheck (parseUrl url)
 
-type ClientM = ExceptT ServantError IO
+-- previously: 
+--type ClientM = ExceptT ServantError IO
```
You can remove this. The place for such notes is in the changelog.

jkarni commented on servant-client/src/Servant/Common/Req.hs:
```diff
@@ -149,20 +152,39 @@ parseRequest url = liftM disableStatusCheck (parseUrl url)
+  , baseUrl :: BaseUrl
+  }
+
+newtype ClientM a = ClientM { runClientM' :: ReaderT ClientEnv (ExceptT ServantError IO) a }
```
A haddock here and above is important. Something like:

```
-- | @ClientM@ is the monad in which client functions run. Contains the 'Manager' and 'BaseUrl' used for requests in the reader environment.
```

jkarni commented on servant-client/src/Servant/Common/Req.hs:
```diff
@@ -149,20 +152,39 @@ parseRequest url = liftM disableStatusCheck (parseUrl url)
+
+newtype ClientM a = ClientM { runClientM' :: ReaderT ClientEnv (ExceptT ServantError IO) a }
+                    deriving ( Functor, Applicative, Monad, MonadIO
+                             , MonadReader ClientEnv
```
`deriving Generic` too

jkarni commented on servant-client/src/Servant/Common/Req.hs:
```diff
@@ -149,20 +152,39 @@ parseRequest url = liftM disableStatusCheck (parseUrl url)
+  = ClientEnv
+  { manager :: Manager
+  , baseUrl :: BaseUrl
+  }
```
Instance [`HasHttpManager`](https://hackage.haskell.org/package/http-client-0.5.3.1/docs/Network-HTTP-Client.html#t:HasHttpManager).

jkarni replied:
It may be interesting to provide convenience functions:

```
defaultLocalHostEnv :: Port -> IO ClientEnv
defaultLocalHostEnv port = do
   mgr <- newManager defautlManagerSettings
   return $ ClientEnv mgr (BaseUrl Http "localhost" port "")

defaultHttpsClient :: String -> IO ClientEnv
defaultHttpsClient host = ...
```

jkarni commented:
This is a great change even aside from the fact that it'll make using `enter` possible! Thanks!

arianvp commented:
Okay so the following problem occurs:

We need `mtl` for `MonadReader`.  But the version of `mtl` in  `lts-2.22` which we use to test `GHC-7.8.4`   does not yet have `Control.Monad.Except`, whilst the `transformers` package in `lts-2.22` already has `Control.Monad.Trans.Except`.  Sadly enough `lts-2.22` was the latest LTS supporting `7.8.4` so we need to consider building this in a different way I guess?

I'm sure there is a valid build plan for 7.8.4.


An option would be:

Manually add a version of `mtl` that does have `Control.Monad.Except` to the  `stack-7.8.4.yaml` file.

jkarni commented:
Why do we need `Control.Monad.Except`? Won't `Control.Monad.Trans.Except` do?

jkarni commented:
Also, see [mtl-compat](https://hackage.haskell.org/package/mtl-compat-0.2.1.2/docs/Control-Monad-Except.html)

chkl commented:
First of all thank you very much for your feedback; this is very valuable to me.
I'm having a little trouble to understand the relationship between `mtl` and `transfomers`. First of all just adding `mtl-compat` to the cabal-file makes it compile. But I guess having less dependencies is always a good thing so tried to work without `mtl`.

The reason that used `mtl` was that I wanted to give `ClientM` a `MonadError` instance and I think that that class is only defined in `mtl`.  @jkarni : You asked for an instance of MonadError, do we need really need one? Otherwise I would remove that class and revert the other change where I replaced `throwE` with `throwError`. This way we wouldn't need `mtl` nor `mtl-compat`.

jkarni commented:
> I'm having a little trouble to understand the relationship between mtl and transfomers. 

`mtl` provides a bunch of typeclasses, usually describing effects. `transformers` provides monad transformers, and these usually are the primary instances for the `mtl` typeclass. So while `mtl` might have a `MonadError`, it's `transformers` that has the concrete `ExceptT`. (But note that `mtl` re-exports things from `tranformers`!). 

Now, the problem isn't that `MonadError` didn't exist previously - only that it existed in `Control.Monad.Error.Class` (from `mtl`), which has been deprecated. So instead of using `mtl-compat`, you could do something like:

```
{-# LANGUAGE CPP #-}
...
#if MIN_VERSION_mtl(2,2,0)
import Control.Monad.Except (MonadError(..))
#else
import Control.Monad.Error.Class (MonadError(..))
#endif
import Control.Monad.Trans.Except

```

I *think* that should work.vBut if it doesn't, I'm also okay with having `mtl-compat`. And I'm definitely okay having `mtl` in the dependencies. I'm sure it transitively already is.

alpmestan commented:
@jkarni isn't what you describe the purpose of `mtl-compat` though? =)

I agree that depending explicitly on `mtl` isn't a problem.

jkarni commented on servant-client/servant-client.cabal:
```diff
@@ -53,6 +53,7 @@ library
     , text                  >= 1.2      && < 1.3
     , transformers          >= 0.3      && < 0.6
     , transformers-compat   >= 0.4      && < 0.6
+    , http-conduit
```
Definitely not worth this dependency for the `HasManager` instance. Sorry! I didn't know it didn't exist previously!

chkl replied:
Ok, I removed it.

jkarni commented:
@chkl great! Last two things - do you mind adding an entry about this change into the servant-client changelog (labeled `BACKWARDS INCOMPATIBLE`), and squashing the commits into one or two?

chkl commented:
I added an entry to the changelog. I squashed all commits on a new branch https://github.com/chkl/servant/tree/clientm. I didn't see a way to edit my pull-request. Do I open a new one?

jkarni commented:
@chkl either that, or force-pushing into this branch.

chkl commented:
I created a new pull request: https://github.com/haskell-servant/servant/pull/599

jkarni commented:
Superseded by #599
